tests/micro-benchmarks/TestWellFoundedRecursion.fst

```diff
@@ -0,0 +1,96 @@
+module TestWellFoundedRecursion
+
+noeq
+type ord =
+  | Z : ord
+  | Succ : ord -> ord
+  | Lim : (nat -> ord) -> ord
+
+let rec sum (x:ord) (y:ord) =
+  match x with
+  | Z -> y
+  | Succ x -> Succ (sum x y)
+  | Lim f -> Lim (fun n -> sum (f n) y)
+
+noeq
+type t =
+  | T0 : bool -> t
+  | T : (nat -> s) -> t
+and s =
+  | S0 : bool -> s
+  | S : (nat -> t) -> s
+
+let rec neg_t (x:t) =
+  match x with
+  | T0 b -> T0 (not b)
+  | T f -> T (fun y -> neg_s (f y))
+
+and neg_s (x:s) =
+  match x with
+  | S0 b -> S0 (not b)
+  | S f -> S (fun y -> neg_t (f y))
+
+
+noeq
+type tree a =
+  | Leaf : data:a -> tree a
+  | Node : children:(nat -> tree a) -> tree a
+
+let rec map #a #b (f:a -> b) (x:tree a)
+  : Tot (tree b)
+  = match x with
+    | Leaf data -> Leaf (f data)
+    | Node ch -> Node (fun n -> map f (ch n))
+
+let codomain_ordering (#a:Type) (x:tree a{Node? x})
+  : Lemma (forall n. Node?.children x n << x)
+  = ()
+
+let test (#a:Type) (g:nat -> tree a) (n:nat)
+  : Lemma (g n << Node g)
+  = //`g n << Node g` is not automatically provable in the SMT encoding
```

Comment:
Is it because we don't have a `HasType (Node g) tree a` term so the outer pattern in the emitted axiom does not trigger?

Reply:
Yes, that's the reason.